tests/conftest.py

```diff
@@ -28,6 +29,17 @@ def pytest_terminal_summary(terminalreporter, exitstatus, config):
             terminalreporter._session.exitstatus = 13
 
 
+@pytest.fixture(autouse=True)
+def capture_wrap():  # pragma: no cover
+    """
+    Workaround for issues with pytest's capture logging
+    https://github.com/pytest-dev/pytest/issues/5502
+    """
+    sys.stderr.close = lambda *args: None
+    sys.stdout.close = lambda *args: None
+    yield
```

Comment:
I'm slightly nervous about this because it's such a sweeping bit of monkeypatching rather than something that's just targeting the databricks/thrift problem. Not that I can think of any specific issues it will cause us.

Looking at the relevant bit of `databricks` [code](https://github.com/databricks/databricks-sql-python/blob/6d62baa2c744c5942fb9563f0cc98c9f0080fe44/src/databricks/sql/client.py#L142-L152) it looks like the warning gets triggered because we've got connnections that are still open at the point they get garbage collected. So maybe we can so more careful lifecycle management. Or have something which sticks all open connections in a global somewhere, and then a fixture which automatically closes them before we exit.

I'd be happy to wave this through for now if it unblocks us, but I'd like to replace it with something more targeted (even if equally gnarly) before too long.

Reply:
I tried putting a [close_all_sessions](https://docs.sqlalchemy.org/en/13/orm/session_api.html?highlight=close_all_sessions#sqlalchemy.orm.session.close_all_sessions) in either a `pytest_sessionfinish()` or a session fixture, but both still end up with this logging error.

Since we only run the databricks tests manually atm, and the logging error might not cause tests to fail anyway, I could just take the workaround out for now.

Reply:
I think `close_all_sessions` is a SQLAlchemy ORM function, rather than a SQLAlchemy Core one — it can be a bit confusing sometimes which features apply to which.

If we're only running them manually then I might be tempted to take it out for now. I might be being overly paranoid, but monkeypatching `sys.stdout` just gives me the heebie jeebies.

It would be worth making an issue describing the problem and the workaround (and a brief note of the things which didn't work) just so the next one of us to hit this has something to go on. And it might be one of those things which Databricks/pytest just fix for us in an update at some point.

Reply:
I've taken out the workaround, and the CI tests are still passing :shrug: